lib/babe/build.go

```diff
@@ -237,6 +237,147 @@ func (b *BlockBuilder) buildBlockExtrinsics(slot Slot, rt runtime.Instance) []*t
 	return included
 }
 
+// CandidateDescriptor is a unique descriptor of the candidate receipt.
+type CandidateDescriptor struct {
+	// The ID of the para this is a candidate for.
+	ParaID uint32
+	// The hash of the relay-chain block this should be executed in
+	// the context of.
+	// NOTE: the fact that the hash includes this value means that code depends
+	// on this for deduplication. Removing this field is likely to break things.
+	RelayParent common.Hash
+	// The collator's relay-chain account ID
+	Collator []byte // CollatorId
+	// Signature on blake2-256 of components of this receipt:
+	// The para ID, the relay parent, and the `pov_hash`.
+	// TODO: I have made a guess that all signatures are []byte
+	Signature []byte // CollatorSignature
+	// The hash of the `pov-block`.
+	PovHash common.Hash
+}
+
+// UpwardMessage is a message from a parachain to its Relay Chain.
+type UpwardMessage []byte
+
+// OutboundHrmpMessage is an HRMP message seen from the perspective of a sender.
+type OutboundHrmpMessage struct {
+	Recipient uint32
+	Data      []byte
+}
+
+// All Vec<u8> in rust have become []byte here
+
+// ValidationCode is Parachain validation code.
+type ValidationCode []byte
+
+// HeadData is Parachain head data included in the chain.
+type HeadData []byte
+
+// CandidateCommitments are Commitments made in a `CandidateReceipt`. Many of these are outputs of validation.
+type CandidateCommitments struct {
+	// Messages destined to be interpreted by the Relay chain itself.
+	UpwardMessages []UpwardMessage
+	// Horizontal messages sent by the parachain.
+	HorizontalMessages []OutboundHrmpMessage
+	// New validation code.
+	NewValidationCode *ValidationCode
+	// The head-data produced as a result of execution.
+	HeadData HeadData
+	// The number of messages processed from the DMQ.
+	ProcessedDownwardMessages uint32
+	// The mark which specifies the block number up to which all inbound HRMP messages are processed.
+	HrmpWatermark uint32
+}
+
+// CommittedCandidateReceipt is a candidate-receipt with commitments directly included.
+type CommittedCandidateReceipt struct {
+	Descriptor  *CandidateDescriptor
+	Commitments *CandidateCommitments
+}
+
+// ValidityAttestation is an implicit or explicit attestation to the validity of a parachain
+// candidate.
+type ValidityAttestation int
```

Comment:
nit to take less space

```suggestion
type ValidityAttestation uint8
```